File: src/main/java/io/vproxy/vfx/manager/task/TaskManager.java
package io.vproxy.vfx.manager.task;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class TaskManager {
    private static final TaskManager instance = new TaskManager();

    private TaskManager() {
    }

    public static TaskManager get() {
        return instance;
    }

    private final ExecutorService threadPool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 4);

    public void execute(Runnable r) {
        threadPool.execute(r);
    }

    public void terminate() {
        threadPool.shutdown();
    }
}

File: src-ts/src/manager/task/TaskManager.ts
export class TaskManager {
  static readonly instance = new TaskManager();

  static get(): TaskManager { return TaskManager.instance; }

  private constructor() {}

  execute(r: () => void): void {
    // Use queueMicrotask for synchronous-finish tasks; for longer work
    // setTimeout(0) yields to the event loop. We pick setTimeout for parity
    // with ExecutorService semantics (runnable submitted, may not run inline).
    setTimeout(r, 0);
  }

  terminate(): void {
    // No-op in the DOM port; setTimeout tasks cannot be cancelled globally
    // without tracking their handles. vfx only calls terminate() on shutdown.
  }
}
